pkg/core/mpt/billet.go

```diff
@@ -132,7 +143,7 @@ func (b *Billet) putIntoExtension(curr *ExtensionNode, path []byte, val Node) (N
 		return nil, err
 	}
 	curr.next = r
-	return curr, nil
+	return b.tryCollapseExtension(curr), nil
```

Comment:
We're returning from some depth here and `tryCollapse` goes into the same depth again, I don't think that's correct. As `Billet` is special it might as well return `(bool, error)` instead of `(Node, error)`, because you absolutely don't care about this `Node`, you don't need it, what you care about though is whether it has all of the data below it and whether it's collapsed. If it is, you might check if the current node can do the same. And if it can, it can collapse and signal to the upper node that it's done.

Reply:
Notice, that `tryCollapse` functions are *not* recursive, so if we're returning from some depth N, then `tryCollapse` goes just to the first-level children (not to the depth N).